File: android/app/src/main/java/com/syw/max/buttonintegration/Model/OfferDetail.java
package com.syw.max.buttonintegration.Model;

import com.usebutton.sdk.impression.VisibleRateType;
import com.usebutton.sdk.impression.CreativeType;

public class OfferDetail {
    private String offerDetailURL;
    private String offerDetailID;
    private double offerDetailRate;
    private VisibleRateType offerDetailType;
    private CreativeType offerCreativeType;

    // Initialization.
    public OfferDetail(String url, String offerID, double rate, boolean rateIsFixed, String creativeType) {
        this.offerDetailURL = url;
        this.offerDetailID = offerID;
        this.offerDetailRate = rate;
        this.offerDetailType = (rateIsFixed) ? VisibleRateType.FIXED : VisibleRateType.PERCENT;
        this.offerCreativeType = this.getCreativeTypeFor(creativeType);
    }

    // Data functions.
    private CreativeType getCreativeTypeFor(String creativeType) {
        switch (creativeType) {
            case "hero":
                return  CreativeType.HERO;
            case "carousel":
                return CreativeType.CAROUSEL;
            case "list":
                return CreativeType.LIST;
            case "grid":
                return CreativeType.GRID;
            case "detail":
                return CreativeType.DETAIL;
        }

        return CreativeType.OTHER;
    }

    // Getters.
    public String getOfferDetailURL() {
        return this.offerDetailURL;
    }

    public String getOfferDetailID() {
        return this.offerDetailID;
    }

    public double getOfferDetailRate() {
        return this.offerDetailRate;
    }

    public VisibleRateType getOfferDetailType() {
        return this.offerDetailType;
    }

    public CreativeType getOfferCreativeType() {
        return this.offerCreativeType;
    }
}

File: ios/ButtonIntegration/OfferDetail.m
#import "OfferDetail.h"

@interface OfferDetail ()

@property (nonatomic, copy) NSString *offerDetailURL;
@property (nonatomic, copy) NSString *offerDetailID;
@property (nonatomic, assign) double offerDetailRate;
@property (nonatomic, assign) BTNVisibleRateType offerDetailType;
@property (nonatomic, assign) BTNCreativeType offerCreativeType;

@end

@implementation OfferDetail

#pragma mark - Creative type functions.
- (BTNCreativeType)getCreativeTypeFor:(NSString *)creativeType {
  if ([creativeType isEqualToString:@"hero"]) {
    return BTNCreativeTypeHero;
  } else if ([creativeType isEqualToString:@"carousel"]) {
    return BTNCreativeTypeCarousel;
  } else if ([creativeType isEqualToString:@"list"]) {
    return BTNCreativeTypeList;
  } else if ([creativeType isEqualToString:@"grid"]) {
    return BTNCreativeTypeGrid;
  } else if ([creativeType isEqualToString:@"detail"]) {
    return BTNCreativeTypeDetail;
  }
  
  return BTNCreativeTypeOther;
}

#pragma mark - Initialization.
- (instancetype)initWithURL:(NSString *)url
                    offerID:(NSString *)offerID
                       rate:(NSUInteger)rate
                rateIsFixed:(BOOL)isFixed
          offerCreativeType:(NSString *)type {
  self = [super init];
  
  if (self != nil) {
    _offerDetailURL = url;
    _offerDetailID = offerID;
    _offerDetailRate = rate;
    _offerDetailType = (isFixed) ? BTNVisibleRateTypeFixed : BTNVisibleRateTypePercent;
    _offerCreativeType = [self getCreativeTypeFor:type];
  }
  
  return self;
}

@end